Prevent IndexOutOfRangeException in RegexInterpreter

rhirano0715 opened the pull request:
This update fixes the IndexOutOfRangeException in RegexInterpreter by enhancing the `TrackPush` and `TrackPush2` methods. The adjustment involves checking the runtrack position before decrementing it, ensuring that it doesn't become negative, which was the root cause of the exception. This prevents potential out-of-range errors when handling large numbers of repetitions in regular expressions.

Fix #62094

rhirano0715 committed:
Prevent IndexOutOfRangeException in RegexInterpreter

This update fixes the IndexOutOfRangeException in RegexInterpreter by enhancing the `TrackPush` and `TrackPush2` methods. The adjustment involves checking the runtrack position before decrementing it, ensuring that it doesn't become negative, which was the root cause of the exception. This prevents potential out-of-range errors when handling large numbers of repetitions in regular expressions.

Fix #62094

stephentoub commented:
Thank you, but I'm not convinced this is the right fix. Why are the numbers becoming negative in the first place? That would seem to suggest a logic error in how the stack is balanced.

rhirano0715 commented:
@stephentoub 

Thanks for your reply.
I think you are right.
I will look into it some more.

rhirano0715 committed:
Changed to call EnsureStorage() unconditionally.

If EnsureStorage() is called unconditionally, the array will be expanded, so the position will never become negative. When the conditions inside EnsureStorage() are true, it might be necessary to expand the array, regardless of the comparison between newpos and codepos.

https://github.com/dotnet/runtime/blob/6ebc8bd86dbc780b2a2a7daf3ab6020f9104f09e/src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.MultipleMatches.Tests.cs#L461-L469

Before the change, in this test case, EnsureStorage() is not called because newpos == codepos == 6 from the first time until an exception occurs.

Fix #62049

rhirano0715 commented:
I made changes because it seemed good for RegexInterpreter's Goto() and Backtrack() to call EnsureStorage() unconditionally.

https://github.com/dotnet/runtime/blob/5c40bb5636b939fb548492fdeb9d501b599ac5f5/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs#L53-L63

https://github.com/dotnet/runtime/blob/5c40bb5636b939fb548492fdeb9d501b599ac5f5/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs#L147-L150

If EnsureStorage() is called unconditionally, the array will be extended, so it seems that the array's position will never become negative. I think there might be a need to extend when the condition inside EnsureStorage() is true.

https://github.com/dotnet/runtime/blob/5c40bb5636b939fb548492fdeb9d501b599ac5f5/src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.MultipleMatches.Tests.cs#L463-L467

In the case of this test case, EnsureStorage() is not called since newpos == codepos == 6 is always true from the first instance until an exception occurs.

stephentoub commented:
> RegexInterpreter's Goto() and Backtrack() to call EnsureStorage() unconditionally

Thanks. Are you able to measure what impact that has on perf? Hopefully the call would be inlined, and we should end up with effectivlely two comparisons instead of one and it shouldn't be measurable. As long as perf looks good, this general form of fix makes sense.

rhirano0715 commented:
@stephentoub 
Thank you for checking.

> Are you able to measure what impact that has on perf? 

I'm sorry. I don't understand how to do it.
Is it something that's written around here?
https://github.com/dotnet/runtime/blob/7740dbd7fdb54e87c6aec5eb7b193bc7d21b68f1/docs/workflow/README.md?plain=1#L89-L92

steveharter commented:
>I'm sorry. I don't understand how to do it.
Is it something that's written around here?

The common ways to verify perf:
- Add or find an existing RegEx benchmark in https://github.com/dotnet/performance/tree/main/src/benchmarks/micro/libraries/System.Text.RegularExpressions and run before\after.
- Run a temporary benchmark such as by using https://github.com/dotnet/BenchmarkDotNet which is also mentioned in
https://github.com/dotnet/performance/blob/main/docs/microbenchmark-design-guidelines.md

rhirano0715 commented:
@stephentoub 
I ran the existing RegEx benchmarks.
It seems to be fine, but what do you think?

// * Summary *

BenchmarkDotNet v0.13.13-nightly.20240311.145, Windows 11 (10.0.22621.3374/22H2/2022Update/SunValley2)
11th Gen Intel Core i7-1185G7 3.00GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK 9.0.100-preview.1.24101.2
  [Host]     : .NET 9.0.0 (9.0.24.8009), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI
  Job-HPUVJJ : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI
  Job-NTZNCJ : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI

PowerPlanMode=00000000-0000-0000-0000-000000000000  IterationTime=250ms  MaxIterationCount=20
MinIterationCount=3  WarmupCount=1

| Method | Job        | Toolchain                                                                                                | Options                     | Mean     | Error    | StdDev  | Median   | Min      | Max      | Ratio | Allocated | Alloc Ratio |
|------- |----------- |--------------------------------------------------------------------------------------------------------- |---------------------------- |---------:|---------:|--------:|---------:|---------:|---------:|------:|----------:|------------:|
| Count  | Job-HPUVJJ | \after\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe  | None                        | 334.8 ms |  3.33 ms | 0.52 ms | 335.0 ms | 334.1 ms | 335.2 ms |  1.00 |1.05 KB |        1.00 |
| Count  | Job-NTZNCJ | \before\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe | None                        | 333.5 ms |  2.18 ms | 0.34 ms | 333.5 ms | 333.0 ms | 333.9 ms |  1.00 |1.05 KB |        1.00 |
|        |            |                                                                                                          |                             |          |          |         |          |          |          |       |        |             |
| Count  | Job-HPUVJJ | \after\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe  | IgnoreCase                  | 800.0 ms |  9.89 ms | 1.53 ms | 799.8 ms | 798.7 ms | 801.8 ms |  0.98 |1.05 KB |        1.00 |
| Count  | Job-NTZNCJ | \before\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe | IgnoreCase                  | 816.1 ms | 13.59 ms | 2.10 ms | 815.7 ms | 814.0 ms | 818.9 ms |  1.00 |1.05 KB |        1.00 |
|        |            |                                                                                                          |                             |          |          |         |          |          |          |       |        |             |
| Count  | Job-HPUVJJ | \after\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe  | Compiled                    | 305.7 ms |  6.07 ms | 1.58 ms | 305.4 ms | 304.0 ms | 308.1 ms |  1.00 |1.05 KB |        1.00 |
| Count  | Job-NTZNCJ | \before\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe | Compiled                    | 305.1 ms |  4.36 ms | 1.13 ms | 304.6 ms | 304.0 ms | 306.4 ms |  1.00 |1.05 KB |        1.00 |
|        |            |                                                                                                          |                             |          |          |         |          |          |          |       |        |             |
| Count  | Job-HPUVJJ | \after\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe  | IgnoreCase, Compiled        | 575.6 ms |  8.54 ms | 3.79 ms | 576.5 ms | 570.7 ms | 580.1 ms |  0.99 |1.05 KB |        1.00 |
| Count  | Job-NTZNCJ | \before\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe | IgnoreCase, Compiled        | 579.1 ms |  8.68 ms | 3.10 ms | 579.0 ms | 574.2 ms | 583.7 ms |  1.00 |1.05 KB |        1.00 |
|        |            |                                                                                                          |                             |          |          |         |          |          |          |       |        |             |
| Count  | Job-HPUVJJ | \after\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe  | NonBacktracking             | 375.4 ms |  6.17 ms | 4.08 ms | 374.6 ms | 371.7 ms | 383.2 ms |  1.00 |1.05 KB |        1.00 |
| Count  | Job-NTZNCJ | \before\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe | NonBacktracking             | 374.3 ms |  6.47 ms | 3.85 ms | 374.8 ms | 367.2 ms | 378.6 ms |  1.00 |1.05 KB |        1.00 |
|        |            |                                                                                                          |                             |          |          |         |          |          |          |       |        |             |
| Count  | Job-HPUVJJ | \after\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe  | IgnoreCase, NonBacktracking | 820.7 ms | 16.30 ms | 5.81 ms | 821.8 ms | 811.2 ms | 828.5 ms |  0.99 |1.05 KB |        1.00 |
| Count  | Job-NTZNCJ | \before\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe | IgnoreCase, NonBacktracking | 827.7 ms | 14.01 ms | 5.00 ms | 826.2 ms | 822.9 ms | 834.5 ms |  1.00 |1.05 KB |        1.00 |

// * Hints *
Outliers
  Perf_Regex_Industry_SliceSlice.Count: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\after\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=3, WarmupCount=1  -> 1 outlier  was  removed, 2 outliers were detected (334.07 ms, 338.34 ms)
  Perf_Regex_Industry_SliceSlice.Count: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\before\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=3, WarmupCount=1 -> 1 outlier  was  removed (340.24 ms)
  Perf_Regex_Industry_SliceSlice.Count: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\before\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=3, WarmupCount=1 -> 1 outlier  was  removed (597.71 ms)
  Perf_Regex_Industry_SliceSlice.Count: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\after\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=3, WarmupCount=1  -> 1 outlier  was  removed (394.62 ms)
  Perf_Regex_Industry_SliceSlice.Count: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\after\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=3, WarmupCount=1  -> 1 outlier  was  detected (811.21 ms)
  Perf_Regex_Industry_SliceSlice.Count: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\before\artifacts\bin\testhost\net9.0-windows-Release-x64\shared\Microsoft.NETCore.App\9.0.0\CoreRun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=3, WarmupCount=1 -> 1 outlier  was  removed (882.08 ms)

// * Legends *
  Options     : Value of the 'Options' parameter
  Mean        : Arithmetic mean of all measurements
  Error       : Half of 99.9% confidence interval
  StdDev      : Standard deviation of all measurements
  Median      : Value separating the higher half of all measurements (50th percentile)
  Min         : Minimum
  Max         : Maximum
  Ratio       : Mean of the ratio distribution ([Current]/[Baseline])
  Allocated   : Allocated memory per single operation (managed only, inclusive, 1KB = 1024B)
  Alloc Ratio : Allocated memory ratio distribution ([Current]/[Baseline])
  1 ms        : 1 Millisecond (0.001 sec)

// * Diagnostic Output - MemoryDiagnoser *


// ***** BenchmarkRunner: End *****
Run time: 00:03:10 (190.47 sec), executed benchmarks: 12

Global total time: 01:05:22 (3922.2 sec), executed benchmarks: 558

danmoseley reviewed (approved):
Thanks

stephentoub commented:
This looks good. Thanks for the fix!
